### ISSUE TITLE
feat: support non-uniform grid in forward() for differentiable optimization

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#33]
## Summary

`forward()` currently raises `ValueError` when `dz_profile`, `dx_profile`, or `dy_profile` is set (api.py L3064-3071):

```python
raise ValueError(
    "forward() does not support non-uniform mesh profiles "
    "(dx_profile / dy_profile / dz_profile). Use run() instead, "
    "or remove the profiles for a uniform-mesh differentiable sim."
)
```

Since `optimize()` and `topology_optimize()` both call `forward()` internally, **non-uniform grid is currently unusable in any gradient-based inverse design workflow**.

## Motivation

- **Memory**: 3D FDTD + `jax.grad` uses 10-30x forward memory. Non-uniform mesh reduces cell count by 3-10x for antenna problems (fine near substrate, coarse in free space), directly mitigating OOM on 24GB GPUs.
- **Paper**: The IEEE TAP paper would benefit from demonstrating that AD gradients remain correct on non-uniform grids — a claim no competing differentiable FDTD paper has made.
- **Consistency**: `run()` already dispatches to `_run_nonuniform()` when profiles are set. Extending `forward()` with the same dispatch would unify the two paths.

## Proposed change

In `_forward_from_materials()` (or a new `_forward_nonuniform()`), add a non-uniform runner path that:
1. Builds `NonUniformGrid` from stored profiles
2. Assembles materials on the non-uniform grid
3. Runs the JAX-traceable forward simulation with non-uniform cell sizes
4. Returns `ForwardResult` (time_series + optional NTFF)

The key requirement is that the forward pass remains **JAX-differentiable** (`jax.grad`-compatible), which means the non-uniform FDTD update equations must use JAX operations (not NumPy).

## Related

- Issue #31 (memory-efficient inverse design)
- IEEE TAP paper experiments: ex1 (DRA), ex3 (patch topology), ex6 (absorber) would all benefit

## Context

Filed from TAP paper experiment development (2026-04-15).
